File: ios/Packages/SimplicityServices/Sources/SimplicityServices/MediaService.swift
import Foundation
import Mockable
import SimplicityApi

/// The backend's transcode states. Generated as a plain `String`, so this is the only place that
/// knows the spellings.
public enum MediaStatus: String, Sendable {
    case pending = "PENDING"
    case processing = "PROCESSING"
    case ready = "READY"
    case failed = "FAILED"
}

public extension MediaAssetResponse {

    /// Nil for a state this app does not know. Better an unrecognised asset than one silently
    /// treated as ready because the comparison fell through.
    var mediaStatus: MediaStatus? {
        status.flatMap(MediaStatus.init(rawValue:))
    }

    var isReady: Bool {
        mediaStatus == .ready
    }

    /// Ready and failed both end polling. Only these two are terminal.
    var isFinished: Bool {
        mediaStatus == .ready || mediaStatus == .failed
    }
}

@Mockable
public protocol MediaService: AnyObject, Sendable {
    func list(orgId: UUID) async throws -> [MediaAssetResponse]

    /// Returns the asset id and a **single** presigned PUT URL for the whole file. There are no
    /// parts, so a failed upload starts again from nothing.
    func register(
        orgId: UUID,
        filename: String,
        contentType: String,
        sizeBytes: Int64
    ) async throws -> UploadTargetResponse

    func markUploaded(orgId: UUID, assetId: UUID) async throws

    /// There is no single-asset endpoint, so this lists and filters. Wasteful for polling a
    /// transcode, but an organisation has tens of assets rather than thousands, and inventing a
    /// client-side cache to avoid it would be worse than the request.
    func asset(orgId: UUID, assetId: UUID) async throws -> MediaAssetResponse?

    /// Sent as a `text/vtt` body rather than presigned: a caption file is kilobytes where a video
    /// is hundreds of megabytes.
    func setCaptions(orgId: UUID, assetId: UUID, vtt: String) async throws
    func removeCaptions(orgId: UUID, assetId: UUID) async throws
    func delete(orgId: UUID, assetId: UUID) async throws
}

File: android/services/src/main/kotlin/io/simplicity/training/services/MediaService.kt
package io.simplicity.training.services

import io.simplicity.training.api.apis.MediaApi
import io.simplicity.training.api.models.MediaAssetResponse
import io.simplicity.training.api.models.RegisterUploadRequest
import io.simplicity.training.api.models.UploadTargetResponse
import java.util.UUID

/**
 * Video upload and its transcoding.
 *
 * There is no endpoint for a single asset, only a list, so [asset] filters. That is worth stating
 * rather than hiding: polling a list to watch one item transcode is the shape the API forces, and
 * a future `GET /media/{id}` should replace it.
 */
interface MediaService {
    suspend fun register(orgId: UUID, filename: String, contentType: String, sizeBytes: Long): UploadTargetResponse
    suspend fun completeUpload(orgId: UUID, assetId: UUID): MediaAssetResponse
    suspend fun list(orgId: UUID): List<MediaAssetResponse>
    suspend fun asset(orgId: UUID, assetId: UUID): MediaAssetResponse?
}

class MediaServiceImpl(private val api: MediaApi) : MediaService {

    override suspend fun register(orgId: UUID, filename: String, contentType: String, sizeBytes: Long) =
        api.registerUpload(
            orgId,
            RegisterUploadRequest(contentType = contentType, filename = filename, sizeBytes = sizeBytes),
        ).unwrap()

    override suspend fun completeUpload(orgId: UUID, assetId: UUID) =
        api.completeUpload(orgId, assetId).unwrap()

    override suspend fun list(orgId: UUID) = api.listMedia(orgId).unwrap()

    override suspend fun asset(orgId: UUID, assetId: UUID) =
        list(orgId).firstOrNull { it.assetId == assetId }
}

/** The states the server reports. A string on the wire, so the client spells them out once. */
object MediaStatus {
    const val UPLOADING = "UPLOADING"
    const val PROCESSING = "PROCESSING"
    const val READY = "READY"
    const val FAILED = "FAILED"
}
